storage/timestamp_cache.go

```diff
@@ -114,61 +120,151 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
-		// entries or return without adding this entry if necessary.
-		key := tc.cache.MakeKey(start, end)
-		for _, o := range tc.cache.GetOverlaps(key.Start, key.End) {
-			ce := o.Value.(*cacheValue)
-			if ce.readOnly != readOnly {
-				continue
-			}
-			if o.Key.Contains(key) {
-				if !ce.timestamp.Less(timestamp) {
-					return // don't add this key; there's already a cache entry with >= timestamp.
-				}
-				if key.Contains(*o.Key) {
-					// The keys are equal, update the existing cache entry.
-					*ce = cacheValue{timestamp: timestamp, txnID: txnID, readOnly: readOnly}
+		var rg interval.RangeGroup
+		rStart := interval.Comparable(start)
+		rEnd := interval.Comparable(end)
+		r := interval.Range{Start: rStart, End: rEnd}
+
+		cache := tc.wCache
+		if readOnly {
+			cache = tc.rCache
+		}
+
+		// Check existing, overlapping entries and truncate/split/remove if
+		// superseded and in the past. If existing entries are in the future,
+		// subtract from the range/ranges that need to be added to cache.
+		for _, o := range cache.GetOverlaps(start, end) {
+			cv := o.Value.(*cacheValue)
+			if !timestamp.Less(cv.timestamp) {
+				// The existing interval has a timestamp less than or equal to the new interval.
+				// Compare interval ranges to determine how to modify existing interval.
+				sCmp := rStart.Compare(o.Key.Start)
+				eCmp := rEnd.Compare(o.Key.End)
+				switch {
+				case sCmp == 0 && eCmp == 0:
+					// New and old are equal; replace old with new and avoid the need to insert new.
+					//
+					// New: ------------
+					// Old: ------------
+					//
+					// Res: ------------ (new)
+					*cv = cacheValue{timestamp: timestamp, txnID: txnID}
+					cache.MoveToEnd(o.Entry)
 					return
+				case sCmp <= 0 && eCmp >= 0:
+					// New contains old; delete old.
+					//
+					// New: ------------
+					// Old:   --------
+					//
+					// Res:
+					cache.DelEntry(o.Entry)
+				case sCmp > 0 && eCmp < 0:
+					// Old contains new; split up old into two.
+					//
+					// New:     ----
+					// Old: ------------
+					//
+					// Res: ----    ----
+					oldEnd := o.Key.End
+					o.Key.End = rStart
+
+					key := cache.MakeKey(rEnd, oldEnd)
+					entry := makeCacheEntry(key, *cv)
+					cache.AddEntryAfter(entry, o.Entry)
+				case eCmp >= 0:
+					// Partial overlap; truncate old end.
+					//
+					// New:     --------
+					// Old: --------
+					//
+					// Res: ----
+					o.Key.End = rStart
+				case sCmp <= 0:
+					// Partial overlap; truncate old start.
+					//
+					// New: --------
+					// Old:     --------
+					//
+					// Res:         ----
+					o.Key.Start = rEnd
+				default:
+					panic(fmt.Sprintf("no overlap between %v and %v", o.Key.Range, r))
 				}
-			} else if key.Contains(*o.Key) && !timestamp.Less(ce.timestamp) {
-				tc.cache.Del(o.Key) // delete existing key; this cache entry supersedes.
+			} else {
+				// The existing interval has a timestamp greater than the new interval.
+				if rg == nil {
```

Comment:
Ah yeah, I forgot the ordering of the overlapping intervals from `GetOverlaps` was deterministic and ordered by start key. Just made the change using that assumption and it makes things a good bit cleaner.